### ISSUE TITLE
Implement the EncryptPassword connection parameter

### ISSUE THREAD

**Issue by fogzot** [DataAction/AdoNetCore.AseClient#27]
I can connect to a SAP database using the following command line:
```
isql64 -U<user> -S<app> -X
```
I am trying to convert that to a connection string and _I think_ I can translate `-S` to `ApplicationName` but `-X` is a problem because, apparently, EncryptPassword doesn't do anythong and the login fails.

**Comment by senseibaka** [comment]
Hi @fogzot , the `EncryptPassword` feature is not implemented.

Previously we were undecided about implementing it, hence it is marked with a `?`. I will change that to `TODO`.

**Comment by fogzot** [comment]
Thanks. TODO makes more sense than ?.

**Comment by fogzot** [comment]
BTW, is there any documentation about that part of the protocol? I looked a bit around and found almost nothing except some obscure references to a "proprietary encryption" (old versions) or a "new RSA encryption", but nothing more.

**Comment by senseibaka** [comment]
In terms of documentation, all we have to use is what's listed [here](https://github.com/DataAction/AdoNetCore.AseClient/wiki#suggested-dev-reference-material-for-contributers).

I've been digging around a bit too. `Sybase-tds38-102306.pdf` diagrams the encrypted password login flow, and very briefly mentions cipher suites.

What suite is used depends on which flags the client sends to the server:
* `TDS_MSG_SEC_ENCRYPT` appears to result in the "proprietary encryption" you mentioned
* `TDS_MSG_SEC_ENCRYPT | TDS_MSG_SEC_ENCRYPT2` results in a cipher suite from Appendix A.5 in [RFC2246](https://www.ietf.org/rfc/rfc2246.txt)

After doing a Ribo dump of the SAP driver, it appears there's also a `TDS_MSG_SEC_ENCRYPT3` flag, which is not documented in the PDF. I hear there's a newer version of the PDF (TDS 5.0 Version 3.9), but have been unable to locate a copy 😢

**Comment by MateuszGalasinski** [comment]
Is there any chance that issue can be reopened and resolved ? Unfortunately, we need it in order to connect to our databases, as they accept encrypted password.

**Comment by YulerB** [comment]
We need this property supported to allow us to migrate our applications to .NET core, since the encryption of the password is a requirement in our company.

**Comment by c-j-hughes** [comment]
Been doing some digging on this. There's a handshake where the server and the driver negotiate the encryption mechanism to use. My view is that we shouldn't need to support all of them. 

`TDS_MSG_SEC_ENCRYPT` - let's agree we can't support this. We have no way to know how it works. It's proprietary to SAP.

`TDS_MSG_SEC_ENCRYPT2` - it's going to take some trial and error, but we should be able to figure this one out. The [Extended Encrypted Password documentation](http://ondoc.logand.com/d/2219/pdf#page=65) certainly helps.

`TDS_MSG_SEC_ENCRYPT3` - if it is similar enough to `TDS_MSG_SEC_ENCRYPT2` we should try to support this. And given it's number 3, it's probably going to be more secure and will be supported for longer by the servers.

**Comment by senseibaka** [comment]
Here's a shortened ribo output of what happens between the reference driver and the server during login:
```
client >>>>>>>>>>>>> server
  Login Record:
    Password: 0 length string
    Remote Password: 0 length string
    Secure Login Flags: SEC_LOG_ENCRYPT + SEC_LOG_ENCRYPT2 + SEC_LOG_ENCRYPT3 (0xA1)
  
client <<<<<<<<<<<<< server
  LOGINACK
    Status: LOG_NEGOTIATE
  MSG
    Status: MSG_HASARGS (0x01)
    MsgId:  MSG_SEC_ENCRYPT3 (0x001E)
  PARAMFMT
    # Params: 3
    Param 1: INT4
    Param 2: LONGBINARY
    Param 3: LONGBINARY
  PARAMS:
    Value 1: 1
    Value 2: RSA Public Key
    Value 3: Changes with each login, appears to be some sort of nonce
  DONE
    Status: DONE_FINAL
    TranState: TDS_TRAN_PROGRESS

client >>>>>>>>>>>>> server
  MSG
    Status: MSG_HASARGS
    MsgId:  MSG_SEC_LOGPWD3 (0x001F)
  PARAMFMT
    # Params: 1
    Param 1: LONGBINARY
  PARAMS
    Value 1: Encrypted Password Value
  MSG
    Status: MSG_HASARGS
    MsgId:  MSG_SEC_REMPWD3 (0x0020)
  PARAMFMT
    # Params: 2
    Param 1: VARCHAR
    Param 2: LONGBINARY
  PARAMS
    Value 1: null
    Value 2: Encrypted Password Value (Exactly the same as the one for LOGPWD3)

client <<<<<<<<<<<<< server
//back to normal end of login response
  a bunch of ENVCHANGE tokens...
  a bunch of EED tokens...
  LOGINACK
    Status: LOG_SUCCEED
  CAPABILITY
  DONE
```

The driver advertises what schemes it supports, and the server chooses which to use.

I'm not sure which'd be easier to figure out. `2` has fewer parameters, but `3` is what the reference driver and server are selecting (hence is visible in Ribo, which might make it easier to compare...).

Also, as you say, `3` is a bigger number than `2` so potentially more secure/better supported.

**Comment by c-j-hughes** [comment]
For `TDS_MSG_SEC_ENCRYPT2` and `TDS_MSG_SEC_ENCRYPT3`:

SAP has documented a little of how the ASE JConnect (Java) driver performs password encryption:

> Asymmetric password encryption mechanism uses RSA encryption algorithms to encrypt the password being transmitted.
> 
> To perform this RSA encryption, configure your JRE with a suitable Java Cryptography Extension (JCE) provider. The configured JCE provider should be capable of supporting the “RSA/ECB/OAEPWithSHA1AndMGF1Padding” or “RSA/NONE/OAEPWithSHA1AndMGF1Padding” transformation. 

Source: [Setting Up the Java Cryptography Extension (JCE) Provider](http://infocenter.sybase.com/help/index.jsp?topic=/com.sybase.infocenter.dc39001.1600/doc/html/har1356663081132.html)

The `ECB` or `NONE` part of the transformations relates to symmetric encryption and is irrelevant to RSA which uses asymmetric keys. So we need to configure the .NET Core RSA Provider as `OAEPWithSHA1AndMGF1Padding`.

Might take a while to work out how to use the nonce. Could be appended to the password, and encrypted, or XORd with the encrypted password, or maybe one of the [many other RSA parameters](https://docs.microsoft.com/en-us/dotnet/api/system.security.cryptography.rsaparameters?view=netframework-4.7.2).

**Comment by senseibaka** [comment]
Fun fact from starting to work on this: you can SIGSEGV the database process by sending incorrect bits in the `loldsecure` part of the login packet 😜

**Comment by senseibaka** [comment]
Hi @YulerB,

I've taken a crack at this and managed to implement *partial* support for `EncryptPassword`. If you'd like to try it out, you can clone and build [this fork/branch](https://github.com/senseibaka/AdoNetCore.AseClient/tree/encrypt-password). It still needs a bit of cleaning up, so not ready for NuGet yet.

Why did I say *partial*? A couple of reasons...

**The `EncryptPassword` connection string property technically has three modes of operation, but we won't support the third:**

* `0` - Disabled
    * **Supported**
* `1` - Enabled
    * **Supported**
* `2` - Enabled if supported by server, else Disabled
    * **Re-interpreted as `1`** - behaviour will be the same as if the value were `1`
    * In our experience, ASE 15 and 16 support password encryption out of the box so it is pointless to implement discoverability. In addition, the driver user should know whether their server supports encryption or not (and should therefore select `0` or `1` as appropriate).
  
**There seem to exist multiple password encryption schemes/protocols, but we wont' support them all:**

0. No encryption
    * **Supported** - this is what we had to start with
1. Proprietary encryption
    * **Not supported** - no way to infer the implementation
2. RSA-based encryption - not supported
    * **Not supported** - this is documented (the documentation is ambiguous), but not observable in interactions between the reference driver and the server (because it's been superceded by **3**).
    * The lack of observability makes it rather difficult to replicate.
3. RSA-based encryption with a nonce
    * **Supported** - this is observable in the interactions between the reference driver and the server. Really helpful.
    * Works with ASE 15 and 16, which are the versions of ASE we aim to support

Will we support schemes 1 and 2?
TL;DR is no.
ASE 15 and 16 are the versions we aim to support.
As far as we can tell, there's no way to force the use of the older encryption schemes (1 and 2).
We could not observe schemes 1 and 2 in action when talking to ASE 15 and 16.

Therefore, it stands to reason that we shouldn't bother supporting schemes 1 and 2.

**Comment by senseibaka** [comment]
Hi @fogzot , @MateuszGalasinski and @YulerB 

FYI Password Encryption has been released in version 0.14.0 on nuget.

Cheers

**Comment by fogzot** [comment]
@senseibaka that's great and thank you very much for the heads up! I'll try it in the next few days and, hopefully, replace some Java with C#. :D

**Comment by anishravindran** [comment]
Thanks , I was using 0.11.0 version and tried to connect to Sybase using encryptpassword option; it was giving "adaptive server requires encryption of the login password on the network." error. @senseibaka : Thanks , helped the comment. updated to latest version and it started working for ASP.NETCore/Cloud stack. Thanks again.
